src/mavm.rs

```diff
@@ -275,12 +275,317 @@ impl fmt::Display for CodePt {
     }
 }
 
-#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub enum BufferElem {
+    Leaf(Rc<RefCell<Vec<u8>>>),
+    Node(Rc<RefCell<Vec<Buffer>>>, u8),
+    Sparse(Rc<RefCell<Vec<usize>>>, Rc<RefCell<Vec<u8>>>, u8),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Buffer {
+    elem: BufferElem,
+    hash: RefCell<Option<Packed>>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Packed {
+    hash: Uint256,
+    size: usize, // total size
+    packed: usize, // packed levels
+}
+
+fn calc_len(h : u8) -> usize {
+    if h == 0 {
+        return 1024;
+    }
+    return 128*calc_len(h-1);
+}
+
+fn hash_buf(buf: &[u8]) -> Packed {
+    if buf.len() == 0 {
+        return zero_packed(1024);
+    }
+    if buf.len() == 32 {
+        return normal(Uint256::from_bytes(buf).avm_hash(), 32);
+    }
+    let len = buf.len();
+    let h1 = hash_buf(&buf[0..len/2]);
+    let h2 = hash_buf(&buf[len/2 .. len]);
+    if is_zero_hash(&h2) {
+        return pack(&h1);
+    }
+    return normal(Uint256::avm_hash2(&unpack(&h1), &unpack(&h2)), buf.len());
+}
+
+pub fn hash_buffer(buf: &[u8]) -> Uint256 {
+    return unpack(&hash_buf(buf));
+}
+
+fn hash_node(buf: &mut [Buffer], sz: usize) -> Packed {
+    if buf.len() == 1 {
+        return buf[0].hash();
+    }
+    let len = buf.len();
+    let h1 = hash_node(&mut buf[0..len/2], sz/2);
+    let h2 = hash_node(&mut buf[len/2 .. len], sz/2);
+    if is_zero_hash(&h2) {
+        return pack(&h1);
+    }
+    return normal(Uint256::avm_hash2(&unpack(&h1), &unpack(&h2)), sz);
+}
+
+pub fn zero_hash(sz: usize) -> Uint256 {
+    if sz == 32 {
+        return Uint256::zero().avm_hash();
+    }
+    let h1 = zero_hash(sz/2);
+    return Uint256::avm_hash2(&h1, &h1);
+}
+
+fn normal(hash: Uint256, sz: usize) -> Packed {
+    return Packed{size: sz, packed: 0, hash: hash};
+}
+
+fn pack(packed: &Packed) -> Packed {
+    return Packed{size: packed.size, packed: packed.packed+1, hash: packed.hash.clone()}
+}
+
+fn is_zero_hash(packed: &Packed) -> bool {
+    return packed.hash == Uint256::zero().avm_hash();
+}
+
+fn unpack(packed: &Packed) -> Uint256 {
+    let mut res = packed.hash.clone();
+    let mut sz = packed.size;
+    for _i in 0..packed.packed {
+        res = Uint256::avm_hash2(&res, &zero_hash(sz));
+        sz = sz*2;
+    }
+    return res;
+}
+
+fn zero_packed(sz: usize) -> Packed {
+    if sz == 32 {
+        return normal(zero_hash(32), 32);
+    }
+    return pack(&zero_packed(sz/2));
+}
+
+fn hash_sparse(idx: &[usize], buf: &[u8], sz: usize) -> Packed {
+    if sz == 32 {
+        let mut res = [0u8; 32];
+        for i in 0..idx.len() {
```

Comment:
I would look into using range based for loops.  There are a number of areas where they could potentially be applied.

Reply:
I just noticed that this comment was worded incorrectly.  I meant to say non-range based for loops, i.e. loops using iterators over a collection.